src/native/eventpipe/ds-eventpipe-protocol.c

```diff
@@ -415,83 +452,36 @@ eventpipe_protocol_helper_stop_tracing (
 static
 bool
 eventpipe_protocol_helper_collect_tracing (
-	DiagnosticsIpcMessage *message,
-	DiagnosticsIpcStream *stream)
-{
-	ep_return_false_if_nok (message != NULL && stream != NULL);
-
-	bool result = false;
-	EventPipeCollectTracingCommandPayload *payload;
-	payload = (EventPipeCollectTracingCommandPayload *)ds_ipc_message_try_parse_payload (message, eventpipe_collect_tracing_command_try_parse_payload);
-
-	if (!payload) {
-		ds_ipc_message_send_error (stream, DS_IPC_E_BAD_ENCODING);
-		ep_raise_error ();
-	}
-
-	EventPipeSessionID session_id;
-	session_id = ep_enable (
-		NULL,
-		payload->circular_buffer_size_in_mb,
-		dn_vector_data_t (payload->provider_configs, EventPipeProviderConfiguration),
-		dn_vector_size (payload->provider_configs),
-		EP_SESSION_TYPE_IPCSTREAM,
-		payload->serialization_format,
-		true,
-		ds_ipc_stream_get_stream_ref (stream),
-		NULL,
-		NULL);
-
-	if (session_id == 0) {
-		ds_ipc_message_send_error (stream, DS_IPC_E_FAIL);
-		ep_raise_error ();
-	} else {
-		eventpipe_protocol_helper_send_start_tracing_success (stream, session_id);
-		ep_start_streaming (session_id);
-	}
-
-	result = true;
-
-ep_on_exit:
-	ds_eventpipe_collect_tracing_command_payload_free (payload);
-	return result;
-
-ep_on_error:
-	EP_ASSERT (!result);
-	ds_ipc_stream_free (stream);
-	ep_exit_error_handler ();
-}
-
-static
-bool
-eventpipe_protocol_helper_collect_tracing_2 (
-	DiagnosticsIpcMessage *message,
+	EventPipeCollectTracingCommandPayload *payload,
 	DiagnosticsIpcStream *stream)
 {
-	ep_return_false_if_nok (message != NULL && stream != NULL);
+	ep_return_false_if_nok (stream != NULL);
 
+	EventPipeSessionID session_id = 0;
 	bool result = false;
-	EventPipeCollectTracing2CommandPayload *payload;
-	payload = (EventPipeCollectTracing2CommandPayload *)ds_ipc_message_try_parse_payload (message, eventpipe_collect_tracing2_command_try_parse_payload);
+	EventPipeSessionOptions options;
 
 	if (!payload) {
 		ds_ipc_message_send_error (stream, DS_IPC_E_BAD_ENCODING);
 		ep_raise_error ();
```

Comment:
@lateralusX with the addition of the fini call if this line is triggered `ep_session_options_fini` will be called without a prior call to `ep_session_options_init`. I'm not sure how to overcome this issue as I need this check before initializing the sturct. I can't declare `options` below this if as the compiler doesn't know about this variable in `ep_on_exit`.

Is this ok?

Reply:
What I do in a case like this is to move the error handling of input parameter that doesn't need any cleanup to the top of the function and then do a `return false;`

So something like this:

```
...

ep_return_false_if_nok (stream != NULL);

if (!payload) {
    ds_ipc_message_send_error (stream, DS_IPC_E_BAD_ENCODING);
    return false;
}

EventPipeSessionOptions options;
ep_session_options_init(
		&options,
...
```